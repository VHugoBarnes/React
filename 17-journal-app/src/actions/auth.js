import { types } from "../types/types";
import { googleAuthProvider, firebase } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch( startLoading() );
        // auth()
        firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
                dispatch( finishLoading() );
            }).catch( e => {
                console.log(e);
                dispatch( finishLoading() );
            })
    }
};

export const startRegisterWithEmailPasswordName = ( username, email, password ) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) => {
                await user.updateProfile({ displayName: username });
                console.log(user);
                // dispatch(
                //     login(user.uid, user.displayName)
                // );
            }).catch( e => {
                console.log(e);
            });
    }
};

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
            });
    }
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const startLogout = () => {
    return async(dispatch) => {
        // Hacemos logout desde firebase
        await firebase.auth().signOut();
        // Borramos el uid de la memoria de la pÃ¡gina
        dispatch(logout());
    }
};

export const logout = () => ({
        type: types.logout
})

