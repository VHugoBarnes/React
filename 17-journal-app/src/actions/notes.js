import Swal from 'sweetalert2';
import { db } from "../firebase/firebaseConfig";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { fileUpload } from '../helpers/fileUpload';

// Preset name for cloudinary: react-journal

export const startNewNote = () => {

    // Con el getState podemos acceder al estado actual de los reducers
    return async(dispatch, getState) => {
        // Sacamos del store el reducer auth y su valor uid.
        const { uid } = getState().auth;
        // Creamos el objeto
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        }
        // Llamamos a un path en firestore para guardar la información en una colección
        const doc = await db.collection(`${uid}/journal/notes`).add( newNote );
        dispatch( activeNote(doc.id, newNote) );
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes,
})

export const startSaveNote = ( note ) => {
    return async(dispatch, getState) => {
        // Sacamos del store el reducer auth y su valor uid.
        const { uid } = getState().auth;

        if(!note.url) {
            delete note.url;
        }

        // Extraemos la nota (una copia) a una variable
        const noteToFirestore = {...note};
        // Se elimina el id porque no se puee actualizar
        delete noteToFirestore.id;

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    }
}

export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id, 
        note: {id, ...note}
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState) => {
        const {active: activeNote} = getState().notes;
        // console.log(file);
        // console.log(activeNote);
        const fileUrl = await fileUpload(file);
        console.log(fileUrl);
    }
}
