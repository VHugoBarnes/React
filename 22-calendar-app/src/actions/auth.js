import Swal from "sweetalert2";
import { fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";

/**
 * Acción asíncrona que hace una petición al backend
 * para iniciar sesión con los datos mandados por parámetro.
 * 
 * Ocupa thunk, por eso retorna una función (asíncrona).
 * 
 * @param {string} email el correo pasado del formulario
 * @param {string} password la contraseña pasada por el formulario
 */
export const startLogin = ( email, password ) => {
    return async( dispatch ) => {
        
        // Llamada a la API
        const resp = await fetchSinToken( 'auth', { email, password }, 'POST' );

        // La respuesta de la API
        const body = await resp.json();
        
        // La API manda un objeto con la propiedad 'ok'
        if ( body.ok ) {
            // Se almacena en el localstorage el JWT generado.
            // Como no es información sensible lo podemos guardar ahí sin problems ;)
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            // Guardar en el store
            dispatch( login({
                uid: body.uid,
                body: body.name
            }) )
        } else { // En caso de errores :o
            Swal.fire('Error', body.msg, 'error');
        }

    }
}

export const startRegister = ( email, password, name ) => {
    return async( dispatch ) => {
        // Llamada a la API
        const resp = await fetchSinToken( 'auth/new', { email, password, name }, 'POST' );

        // La respuesta de la API
        const body = await resp.json();
        
        // La API manda un objeto con la propiedad 'ok'
        if ( body.ok ) {
            // Se almacena en el localstorage el JWT generado.
            // Como no es información sensible lo podemos guardar ahí sin problems ;)
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            // Guardar en el store
            dispatch( login({
                uid: body.uid,
                body: body.name
            }) )
        } else { // En caso de errores :o
            Swal.fire('Error', body.msg, 'error');
        }
    }
}

/**
 * Acción síncrona para hacer un login en la web.
 * Básicamente lo que hace es guardar en el store el usuario pasado por param.
 * 
 * @param {object} user El objeto que contiene la información del usuario para almacenarlo en el store
 */
const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});