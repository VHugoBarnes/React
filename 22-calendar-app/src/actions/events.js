import Swal from 'sweetalert2';
import { fetchConToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';
import { types } from '../types/types';

/**
 * 
 * Acción asíncrona que dispara una petición POST
 * al servidor para guardar eventos en la base de datos.
 * 
 * @param {object} event objeto con los datos del evento
 */
export const eventStartAddNew = ( event ) => {
    return async( dispatch, getState ) => {

        // Obtenemos el uid y el name de redux.
        const { uid, name } = getState().auth;
        
        try {
            // Realizar petición POST al server
            const resp = await fetchConToken('events', event, 'POST');
            // Almacenamos en body el objeto que devuelve el server
            const body = await resp.json();
            
            // Si se guardó correctamente:
            if ( body.ok ) {
                // Guardar en el store propiedades
                // en event
                event.id = body.evento.id;
                event.user = {
                    _id: uid,
                    name: name
                }
                dispatch( eventAddNew( event ) );
            }

        } catch (error) {
            console.error(error);
        }

    }
}

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({type: types.eventClearActiveEvent});

export const eventStartUpdate = ( event ) => {
    return async( dispatch ) => {

        try {
            
            const resp = await fetchConToken(`events/${ event.id }`, event, 'PUT');
            const body = await resp.json();

            if ( body.ok ) {
                dispatch( eventUpdated(event) );
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.error(error);
        }

    }
}

const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {
    return async( dispatch, getState ) => {

        const { id } = getState().calendar.activeEvent;

        try {
            
            const resp = await fetchConToken(`events/${ id }`, {}, 'DELETE');
            const body = await resp.json();

            if ( body.ok ) {
                dispatch( eventDeleted() );
            } else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.error(error);
        }

    }
}

export const eventDeleted = () => ({type: types.eventDeleted});

export const eventStartLoading = () => {
    return async( dispatch ) => {

        try {
            
            const resp = await fetchConToken('events');
            const body = await resp.json();

            const events = prepareEvents( body.eventos );
            dispatch( eventLoaded(events) );

        } catch (error) {
            console.error(error);
        }

    }
}

const eventLoaded = ( events ) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({ type: types.eventLogout });
