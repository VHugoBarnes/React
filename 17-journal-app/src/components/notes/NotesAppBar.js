import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import {startSaveNote} from '../../actions/notes';

export const NotesAppBar = () => {
    const {active} = useSelector(state => state.notes);
    const { date } = active;
    const noteDate = moment(date);
    const dispatch = useDispatch();

    const handleSubmit = () => {
        dispatch(startSaveNote(active));
    }

    return (
        <div className="notes__appbar">
            <span>{noteDate.format('MMMM Do YYYY')}</span>

            <div>
                <button className="btn">
                    Picture
                </button>
                <button 
                    className="btn"
                    onClick={handleSubmit}
                >
                    Save
                </button>
            </div>
        </div>
    );
};
