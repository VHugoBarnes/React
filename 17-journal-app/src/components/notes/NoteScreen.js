import React from 'react';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {
    return (
        <div className="notes__main-content">
            <NotesAppBar />
            <div className="notes__content">
                <input 
                    type="text"
                    placeholder="Title"
                    className="notes__title-input"
                    autoComplete="off"
                />

                <textarea 
                    placeholder="What happened today?"
                    className="notes__text-area"
                    autoComplete="off"
                >
                </textarea>
                <div className="notes__image">
                    <img 
                        src="https://c.files.bbci.co.uk/92F7/production/_101232673_angry.jpg"
                        alt="image"
                    />
                </div>
            </div>
        </div>
    );
};
