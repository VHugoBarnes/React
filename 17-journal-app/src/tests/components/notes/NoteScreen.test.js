import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
// LibrerÃ­as para hacer test con thunk
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote, startDeleting } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn(),
    startDeleting: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: 1234,
            title: 'Hola',
            body: 'mundo',
            date: 0
        },
        notes: []
    }
};
// El estado del store en este instante
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <NoteScreen />
    </Provider>
);

describe('Pruebas en el <NoteScreen />', () => {
   
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('Debe de disparar el activeNote', () => {
        wrapper.find('input[name="title"]').simulate('change',{
            target: {
                name: 'title',
                value: 'Hola de nuevo'
            }
        });

        expect(activeNote).toHaveBeenLastCalledWith(1234, {
            body: 'mundo',
            date: 0,
            id: 1234,
            title: 'Hola de nuevo'
        })
    });
    
});
