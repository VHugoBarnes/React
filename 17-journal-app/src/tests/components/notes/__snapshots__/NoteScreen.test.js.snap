// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <NoteScreen /> Debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn animate__faster"
>
  <div
    className="notes__appbar"
  >
    <span>
      December 31st 1969
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value="Hola"
    />
    <textarea
      autoComplete="off"
      className="notes__text-area"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
